### PR TITLE
Migrate merge windows rule to iterative optimizer

### PR DESCRIPTION
**Pull request by amrutagokhale**
The last commit removes the MergeWindows optimizer, in favor of the new MergeAdjacentWindows optimizer rule. This commit can be merged a bit later, after the first three commits go in.

TD review here: https://github.com/Teradata/presto/pull/495

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+        tester.assertThat(new MergeAdjacentWindows())
+                .on(p ->
+                        p.window(
+                                newWindowNodeSpecification(p, "a"),
+                                ImmutableMap.of(p.symbol("avg_1", BIGINT), newWindowNodeFunction("avg", windowA, "a")),
+                                p.window(
+                                        newWindowNodeSpecification(p, "a"),
+                                        ImmutableMap.of(p.symbol("sum_1", BIGINT), newWindowNodeFunction("sum", windowA, "a")),
+                                        p.values(p.symbol("a", BIGINT), p.symbol("b", BIGINT))
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java]: You don't need symbol `b` here.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                                functionCall("avg", Optional.empty(), ImmutableList.of(columnAAlias)),
+                                functionCall("sum", Optional.empty(), ImmutableList.of(columnAAlias))),
+                        values(ImmutableMap.of(columnAAlias, 0, columnBAlias, 1))));
+    }
+
+    private static WindowNode.Specification newWindowNodeSpecification(PlanBuilder planBuilder, String symbolName)
+    {
+        return new WindowNode.Specification(ImmutableList.of(planBuilder.symbol(symbolName, BIGINT)), ImmutableList.of(), ImmutableMap.of());
+    }
+
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java]: Add a test for the same `partition by` but different ordering, e.g. `window(partition by A, order by B ASC)` preceded by `window(partition by A, order by B DESC)`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java [hunk 3]
```diff
@@ -66,28 +66,71 @@ private static boolean dependsOn(WindowNode parent, WindowNode child)
...
+        if (comparison != 0) {
+            return comparison;
+        }
+
+        comparison = compareOrderBy(o1, o2);
+        if (comparison != 0) {
+            return comparison;
+        }
+
+        // If PartitionBy and OrderBy clauses are identical, let' establish an arbitrary order to prevent non-deterministic results of swapping WindowNodes in such a case
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java]: typo: `let'`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java [hunk 4]
```diff
@@ -66,28 +66,71 @@ private static boolean dependsOn(WindowNode parent, WindowNode child)
...
         Iterator<Symbol> iterator2 = o2.getPartitionBy().iterator();
 
         while (iterator1.hasNext() && iterator2.hasNext()) {
             Symbol symbol1 = iterator1.next();
             Symbol symbol2 = iterator2.next();
 
-            int comparison = symbol1.compareTo(symbol2);
-            if (comparison != 0) {
-                return comparison;
+            int partitionByComparison = symbol1.compareTo(symbol2);
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java]: Unnecessary change.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 5]
```diff
@@ -173,7 +174,16 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                         stats,
                         ImmutableSet.of(new SimplifyCountOverConstant())),
                 new WindowFilterPushDown(metadata), // This must run after PredicatePushDown and LimitPushDown so that it squashes any successive filter nodes and limits
-                new MergeWindows(),
+                new IterativeOptimizer(
+                                stats,
+                                ImmutableList.of(
+                                        new UnaliasSymbolReferences(),
+                                        new MergeWindows()),
+                                ImmutableSet.of(
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Add a `TODO` to add a `UnaliasSymbolReferences` once it's ported to the new iterative optimizer.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.window;
+import static com.facebook.presto.sql.tree.FrameBound.Type.CURRENT_ROW;
+import static com.facebook.presto.sql.tree.FrameBound.Type.UNBOUNDED_PRECEDING;
+
+public class TestMergeAdjacentWindows
+{
+    private final RuleTester tester = new RuleTester();
+
+    private WindowNode.Frame frame;
```
Comment (amrutagokhale) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java]: I can change it. But that's exactly how it is in `TestSwapAdjacentWindowsByPartitionsOrder.java` at present :)

Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java]: Ouch, then it should be changed there too. :)

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java (hunk 2 above)
Comment (amrutagokhale) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java]: There is a test `testNotMergeDifferentNullOrdering` in `TestMergeWindows.java` with identical `partition by` clause, but different `NULLS ordering` in `order by` clause, which would cover similar scenario.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java (hunk 4 above)
Comment (amrutagokhale) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java]: I renamed the `comparison` variable to `partitionByComparison` so as to distinguish this code from the code in `compareOrderBy` for better readability.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 5 above)
Comment (amrutagokhale) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I can do that, but this entire block of code gets replaced in the next commit, where I simply add `MergeAdjacentWindows` to `SwapAdjacentWindowsBySpecifications`, and you have already added a TODO there.

**Comment by amrutagokhale** [comment]
@maciejgrzybek Addressed your comments.

**Comment by maciejgrzybek** [comment]
@martint: ready for you :)

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.window;
+import static com.facebook.presto.sql.tree.FrameBound.Type.CURRENT_ROW;
+import static com.facebook.presto.sql.tree.FrameBound.Type.UNBOUNDED_PRECEDING;
+
+public class TestMergeAdjacentWindows
+{
+    private final RuleTester tester = new RuleTester();
+
+    private static final WindowNode.Frame frame = new WindowNode.Frame(WindowFrame.Type.RANGE, UNBOUNDED_PRECEDING,
+            Optional.empty(), CURRENT_ROW, Optional.empty());;
```
Comment (sanjay990) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java]: Unnecessary semicolon
